File: src/oasees_sdk/stack/demo-server/lib/nwo.js
// NWO - Node Workload Orchestrator
// Receives swarm instructions from DSM, checks resources, dispatches to NCE

const EventEmitter = require('events');
const crypto = require('crypto');

class NWO extends EventEmitter {
  constructor(nce, nds) {
    super();
    this.nce = nce;
    this.nds = nds;
    this.emit('action', { component: 'NWO', action: 'init', log: '[NWO] Initialized. Swarm-coordinated orchestrator ready.' });
  }

  handleSwarmNotification(payload) {
    this.emit('action', {
      component: 'NWO', action: 'receiveSwarm',
      log: `[NWO][M6] ⚡ Swarm Instruction received from DSM — type: ${payload.type}`,
      data: payload
    });

    this.emit('action', {
      component: 'NWO', action: 'swarmContext',
      log: '[NWO] Swarm Intelligence: validating instruction against local node capabilities...',
    });

    const data = payload.data || {};
    const serviceId = data.serviceId;
    const requirements = (data.parameters || {}).requirements;
    const swarmNodes = (data.swarm || {}).nodes || [];

    // 0. Multi-node scheduling
    const schedule = this.scheduleNodes(swarmNodes, requirements);
    const selectedNode = (schedule.selected || {}).node_id;
    if (selectedNode && selectedNode !== this.nds.nodeProfile.node_id) {
      this.emit('action', {
        component: 'NWO', action: 'schedule',
        log: `[NWO] Selected remote node ${selectedNode} — forwarding ticket.`,
        data: schedule
      });
      return { status: 'FORWARDED', target_node: selectedNode, schedule };
    }

    // 1. Check resources via NDS
    const resourcesOk = this.nds.checkResources(requirements);
    if (!resourcesOk) {
      this.emit('action', {
        component: 'NWO', action: 'resourceFailed',
        log: `[NWO] Swarm-coordinated execution ABORTED — insufficient resources for: ${serviceId}`,
        data: { status: 'FAILED', reason: 'Insufficient Resources' }
      });
      return { status: 'FAILED', reason: 'Insufficient Resources' };
    }

    // 2. Create execution ticket
    const jobId = `job-${crypto.randomBytes(4).toString('hex')}`;
    const allocation = this.nds.reserveResources(jobId, requirements);
    const executionStrategy = (data.parameters || {}).execution_strategy || 'local-k8s';

    const ticket = {
      job_id: jobId,
      service_id: serviceId,
      image: (data.parameters || {}).image,
      env_vars: (data.parameters || {}).env,
      volume_mounts: (data.parameters || {}).volumes || null,
      allocation,
      execution_strategy: executionStrategy,
      target_node: selectedNode || this.nds.nodeProfile.node_id,
      schedule
    };

    this.emit('action', {
      component: 'NWO', action: 'ticketCreated',
      log: `[NWO] Swarm execution ticket created: ${jobId}`,
      data: ticket
    });

    // 3. Dispatch to NCE (M7)
    this.emit('action', {
      component: 'NWO', action: 'dispatch',
      log: `[NWO][M7] Dispatching execution ticket to NCE — Swarm-coordinated execution underway`,
      data: { job_id: jobId, service_id: serviceId }
    });

    const result = this.nce.submitTicket(ticket);

    return { status: 'ACCEPTED', job_id: jobId, execution_strategy: executionStrategy, target_node: ticket.target_node, schedule, result };
  }

  scoreNode(node, requirements) {
    let score = 0;
    if (requirements && requirements.gpu) score += node.gpu ? 50 : -100;
    score += Math.max(0, 100 - (node.latency_ms || 100)) / 2;
    score += Math.max(0, (1 - (node.load || 0.5)) * 20);
    score += (node.cpu || 0) * 2;
    const recentJobs = node.recent_jobs || 0;
    score -= recentJobs * 5;
    return Math.round(score * 100) / 100;
  }

  scheduleNodes(candidates, requirements) {
    if (!candidates || !candidates.length) {
      return { selected: { node_id: this.nds.nodeProfile.node_id }, scores: [] };
    }
    const scores = candidates.map(n => ({ ...n, score: this.scoreNode(n, requirements) }));
    scores.sort((a, b) => b.score - a.score);
    return { selected: scores[0], scores };
  }
}

module.exports = NWO;

File: src/oasees_sdk/stack/nwo.py

import uuid
from .nce import NCE
from .nds import NDS

class NWO:
    """
    Node Workload Orchestrator (NWO) - Mock Implementation
    
    Responsible for:
    1. Listening for Swarm Instructions (Interface M6).
    2. Orchestrating local resources via NDS.
    3. Building execution tickets for NCE.
    4. Dispatching execution to NCE.
    """
    
    def __init__(self, nce: NCE, nds: NDS):
        self.nce = nce
        self.nds = nds
        print("[NWO] Initialized.")

    def handle_swarm_notification(self, payload):
        """
        M6: Swarm Execution Trigger
        Receives a CloudEvent/JSON from DSM.
        """
        print(f"[NWO][M6] Received Swarm Notification: {payload.get('type')}")
        data = payload.get("data", {})
        service_id = data.get("serviceId")
        swarm_nodes = (data.get("swarm") or {}).get("nodes", [])
        requirements = data.get("parameters", {}).get("requirements")

        # 1. Multi-node Scheduling (Swarm)
        schedule = self.schedule_nodes(swarm_nodes, requirements)
        selected_node = (schedule.get("selected") or {}).get("node_id")
        if selected_node and selected_node != self.nds.node_profile.get("node_id"):
            print(f"[NWO] Selected remote node {selected_node}. Forwarding ticket.")
            return {"status": "FORWARDED", "target_node": selected_node, "schedule": schedule}

        # 2. Check Resources (NDS)
        if not self.nds.check_resources(requirements):
            print(f"[NWO] Resource check failed for service: {service_id}")
            return {"status": "FAILED", "reason": "Insufficient Resources"}
        
        # 3. Prepare Execution Ticket
        job_id = f"job-{uuid.uuid4().hex[:8]}"
        allocation = self.nds.reserve_resources(job_id, requirements)
        execution_strategy = data.get("parameters", {}).get("execution_strategy", "local-k8s")
        execution_ticket = {
            "job_id": job_id,
            "service_id": service_id,
            "image": data.get("parameters", {}).get("image"),
            "env_vars": data.get("parameters", {}).get("env"),
            "volume_mounts": data.get("parameters", {}).get("volumes"),
            "allocation": allocation,
            "execution_strategy": execution_strategy,
            "target_node": selected_node or self.nds.node_profile.get("node_id"),
            "schedule": schedule
        }
        
        # 4. Dispatch to NCE (M7)
        print(f"[NWO][M7] Dispatching Execution Ticket to NCE")
        self.nce.submit_ticket(execution_ticket)
        
        return {
            "status": "ACCEPTED",
            "job_id": job_id,
            "execution_strategy": execution_strategy,
            "target_node": execution_ticket["target_node"],
            "schedule": schedule
        }

    def score_node(self, node, requirements):
        score = 0.0
        if requirements and requirements.get("gpu"):
            score += 50 if node.get("gpu") else -100
        score += max(0, 100 - node.get("latency_ms", 100)) / 2
        score += max(0, (1 - node.get("load", 0.5)) * 20)
        score += node.get("cpu", 0) * 2

        # Fairness: prefer nodes with fewer recent jobs
        recent_jobs = node.get("recent_jobs", 0)
        score -= recent_jobs * 5
        return round(score, 2)

    def schedule_nodes(self, candidates, requirements):
        if not candidates:
            return {"selected": {"node_id": self.nds.node_profile.get("node_id")}, "scores": []}
        scores = []
        for node in candidates:
            scored = dict(node)
            scored["score"] = self.score_node(node, requirements)
            scores.append(scored)
        scores.sort(key=lambda n: n["score"], reverse=True)
        return {"selected": scores[0], "scores": scores}
